File: 23246.java
import java.util.*;
public class Main {
    class Climber {
        public int back;
        public int r1, r2, r3;
        int prod() {
            return this.r1 * this.r2 * this.r3;
        }
        int sum() {
            return this.r1 + this.r2 + this.r3;
        }
    }
    public boolean less(Climber x, Climber y) {
        return x.prod() < y.prod() ||
            (x.prod() == y.prod() && x.sum() < y.sum()) ||
            (x.prod() == y.prod() && x.sum() == y.sum() && x.back < y.back);
    }
    public void solve() {
        Scanner sc = new Scanner(System.in);
		int n;
		n = sc.nextInt();
        List<Climber> a = new ArrayList<>();
        int i;
        for(i=0;i<n;i++) {
            Climber x = new Climber();
            x.back = sc.nextInt();
            x.r1 = sc.nextInt();
            x.r2 = sc.nextInt();
            x.r3 = sc.nextInt();
            a.add(x);
        }
        int trial;
        for(trial=0;trial<3;trial++) {
            Climber top = a.get(0);
            for (Climber x : a) {
                if (less(x, top)) {
                    top = x;
                }
            }
            System.out.printf("%d ", top.back);
            a.remove(top);
        }
        System.out.printf("\n");
    }
	public static void main(String args[]){
        Main s = new Main();
        s.solve();
	}
}

File: 23246.py
def prod(t):
    _, c1, c2, c3 = t
    return c1 * c2 * c3

def sum(t):
    _, c1, c2, c3 = t
    return c1 + c2 + c3

def less(c1, c2):
    back1 = c1[0]
    back2 = c2[0]
    return prod(c1) < prod(c2) or \
        (prod(c1) == prod(c2) and sum(c1) < sum(c2)) or \
        (prod(c1) == prod(c2) and sum(c1) == sum(c2) and back1 < back2)

n = int(input().strip())
a = []
for i in range(n):
    back, r1, r2, r3 = [int(x.strip()) for x in input().split()]
    a.append((back, r1, r2, r3))

ans = []
for i in range(3):
    top = a[0]
    for c in a:
        if less(c, top):
            top = c
    ans.append(top[0])
    a.remove(top)
print(' '.join([str(x) for x in ans]))
